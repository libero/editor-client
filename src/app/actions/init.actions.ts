import { createAction } from 'app/utils/action.utils';

export const initApplication = createAction<void>('INIT_LIBERO');
