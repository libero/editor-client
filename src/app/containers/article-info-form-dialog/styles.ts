import { makeStyles } from '@material-ui/core/styles';

export const useArticleInfoFormStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(1, 2.5, 0, 2.5)
  },

  inputField: {
    marginBottom: theme.spacing(2.5)
  },

  copyrightStatement: {
    paddingBottom: theme.spacing(2),
    paddingTop: theme.spacing(2)
  },

  buttonPanel: {
    display: 'flex',
    justifyContent: 'flex-end',
    marginTop: theme.spacing(4),
    '& > *:last-child': {
      marginLeft: theme.spacing(2)
    }
  },
  spacer: {
    flexGrow: 1
  }
}));
