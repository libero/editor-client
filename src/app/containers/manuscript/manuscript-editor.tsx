import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { EditorState, Transaction } from 'prosemirror-state';

import { getAbstract, getKeywordGroups, getTitle } from 'app/selectors/manuscript.selectors';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { RichTextEditor } from 'app/components/rich-text-editor';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { KeywordsEditor } from 'app/components/keywords';
import { KeywordGroups } from 'app/models/manuscript';
import { useManuscriptStyles } from './styles';
import { SortableAuthorsList } from './sortable-authors-list';
import { AuthorsInfoDetails } from 'app/containers/manuscript/authors-info-details';
import { AffiliationsList } from 'app/containers/manuscript/affiliations-list';
import { ReferenceList } from 'app/containers/manuscript/references-list';
import { ArticleInformation } from 'app/containers/manuscript/article-information';
import { getFocusedEditorStatePath } from 'app/selectors/manuscript-editor.selectors';

const isInputFocused = (inputName: string, focusedPath?: string) => {
  return Boolean(focusedPath) && focusedPath.startsWith(inputName);
};

export const ManuscriptEditor: React.FC = () => {
  const classes = useManuscriptStyles();
  const dispatch = useDispatch();

  const title: EditorState = useSelector(getTitle);
  const abstract: EditorState = useSelector(getAbstract);
  const focusedPath = useSelector(getFocusedEditorStatePath);
  const allKeywords: KeywordGroups = useSelector(getKeywordGroups);

  const handleTitleChange = useCallback(
    (diff: Transaction) => {
      dispatch(manuscriptActions.updateTitleAction(diff));
    },
    [dispatch]
  );

  const handleKeywordsChange = (keywordGroup: string, index: number, diff: Transaction): void => {
    dispatch(manuscriptActions.updateKeywordAction({ keywordGroup, index, change: diff }));
  };

  const handleNewKeywordChange = (keywordGroup: string, diff: Transaction): void => {
    dispatch(manuscriptActions.updateNewKeywordAction({ keywordGroup, change: diff }));
  };

  const handleAbstractChange = useCallback(
    (diff: Transaction): void => {
      dispatch(manuscriptActions.updateAbstractAction(diff));
    },
    [dispatch]
  );

  const handleKeywordDelete = (keywordGroup: string, index: number): void => {
    dispatch(manuscriptActions.deleteKeywordAction({ keywordGroup, index }));
  };

  const handleKeywordAdd = (keywordGroup: string, keyword: EditorState): void => {
    dispatch(manuscriptActions.addNewKeywordAction({ keywordGroup, keyword }));
  };

  const handleFocusSwitch = useCallback(
    (state: EditorState, path: string) => {
      dispatch(manuscriptEditorActions.setFocusAction(path));
    },
    [dispatch]
  );

  const handleBlur = useCallback(() => {
    dispatch(manuscriptEditorActions.removeFocusAction());
  }, [dispatch]);

  const handleKeywordFocus = (group: string, index: number | undefined, isNewKeywordFocused: boolean): void => {
    const kwdIndexPath = isNewKeywordFocused ? 'newKeyword' : `keywords.${index}`;
    handleFocusSwitch(null, ['keywordGroups', group, kwdIndexPath].join('.'));
  };

  const renderKeywords = (keywordGroups: KeywordGroups): JSX.Element[] => {
    return Object.entries(keywordGroups).map(([groupType, group]) => {
      return (
        <KeywordsEditor
          key={groupType}
          keywords={group.keywords}
          newKeyword={group.newKeyword}
          activeKeywordPath={focusedPath}
          name={groupType}
          label={group.title || 'Keywords'}
          onNewKeywordChange={handleNewKeywordChange}
          onAdd={handleKeywordAdd}
          onChange={handleKeywordsChange}
          onDelete={handleKeywordDelete}
          onFocusSwitch={handleKeywordFocus}
          onBlur={handleBlur}
        />
      );
    });
  };

  return (
    <div className={classes.content}>
      <div aria-hidden="true" className={classes.toolbarPlaceholder} />
      <RichTextEditor
        editorState={title}
        label="Title"
        isActive={isInputFocused('title', focusedPath)}
        name="title"
        onChange={handleTitleChange}
        onFocusSwitch={handleFocusSwitch}
      />
      <SortableAuthorsList />
      <AffiliationsList />
      <RichTextEditor
        editorState={abstract}
        label="Abstract"
        name="abstract"
        isActive={isInputFocused('abstract', focusedPath)}
        onChange={handleAbstractChange}
        onFocusSwitch={handleFocusSwitch}
      />
      {renderKeywords(allKeywords)}
      <ArticleInformation />
      <AuthorsInfoDetails />
      <ReferenceList />
    </div>
  );
};
