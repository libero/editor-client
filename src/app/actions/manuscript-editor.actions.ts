import { createAction } from 'redux-act';
import { TableOfContents, TOCEntry } from 'app/types/manuscript';

export interface ModalPayload<P = {}> {
  component: React.FC<P>;
  props?: P;
  title?: string;
}

export const setFocusAction = createAction<string>('SET_FOCUS');
export const setBodyTOCAction = createAction<TableOfContents>('SET_BODY_TOC');
export const setLastSyncTimestamp = createAction<number>('SET_LAST_SYNC_TIMESTAMP');
export const setLastSyncFailed = createAction<void>('SET_LAST_SYNC_FAILED');
export const setManuscriptId = createAction<string>('SET_MANUSCRIPT_ID');
export const scrollIntoViewAction = createAction<TOCEntry>('SCROLL_INTO_VIEW');
export const removeFocusAction = createAction<void>('REMOVE_FOCUS');
export const updateFocusPathAction = createAction<string>('UPDATE_FOCUS_PATH');
export const showModalDialog = createAction<ModalPayload>('SHOW_MODAL');
export const hideModalDialog = createAction<void>('HIDE_MODAL');
