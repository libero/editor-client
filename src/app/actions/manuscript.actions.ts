import { createAction } from 'redux-act';
import { EditorState, Transaction } from 'prosemirror-state';

import { Manuscript } from 'app/types/manuscript';
import { createAsyncAction } from 'app/utils/action.utils';
import { Person } from 'app/models/person';
import { Affiliation } from 'app/models/affiliation';
import { Reference } from 'app/models/reference';
import { RelatedArticle } from 'app/models/related-article';
import { ArticleInformation } from 'app/models/article-information';
import { ListType } from 'app/types/utility.types';

export interface KeywordUpdatePayload {
  keywordGroup: string;
  index: number;
  change: Transaction;
}

export interface NewKeywordUpdatePayload {
  keywordGroup: string;
  change: Transaction;
}

export interface KeywordAddPayload {
  keywordGroup: string;
  keyword: EditorState;
}

export interface KeywordDeletePayload {
  keywordGroup: string;
  index: number;
}

export interface MoveAuthorPayload {
  index: number;
  author: Person;
}

export interface LinkAffiliationsPayload {
  affiliation: Affiliation;
  authors: Person[];
}

export interface ApplyChangePayload {
  path: string;
  change: Transaction;
}

export const loadManuscriptAction = createAsyncAction<string, Manuscript>('LOAD_MANUSCRIPT');
export const updateTitleAction = createAction<Transaction>('UPDATE_TITLE');
export const updateAbstractAction = createAction<Transaction>('UPDATE_ABSTRACT');
export const updateBodyAction = createAction<Transaction>('UPDATE_BODY');
export const updateAcknowledgementsAction = createAction<Transaction>('UPDATE_ACKNOWLEDGEMENTS');
export const updateImpactStatementAction = createAction<Transaction>('UPDATE_IMPACT_STATEMENT');
export const updateKeywordAction = createAction<KeywordUpdatePayload>('UPDATE_KEYWORD');
export const updateNewKeywordAction = createAction<NewKeywordUpdatePayload>('UPDATE_NEW_KEYWORD');
export const deleteKeywordAction = createAction<KeywordDeletePayload>('DELETE_KEYWORD');
export const addNewKeywordAction = createAction<KeywordAddPayload>('ADD_KEYWORD');
export const updateAuthorAction = createAction<Person>('UPDATE_AUTHOR');
export const updateArticleInformationAction = createAction<ArticleInformation>('UPDATE_ARTICLE_INFORMATION');
export const addAuthorAction = createAction<Person>('ADD_AUTHOR');
export const moveAuthorAction = createAction<MoveAuthorPayload>('MOVE_AUTHOR');
export const deleteAuthorAction = createAction<Person>('DELETE_AUTHOR');
export const updateAffiliationAction = createAction<Affiliation>('UPDATE_AFFILIATION');
export const addAffiliationAction = createAction<Affiliation>('ADD_AFFILIATION');
export const deleteAffiliationAction = createAction<Affiliation>('DELETE_AFFILIATION');
export const linkAffiliationsAction = createAction<LinkAffiliationsPayload>('LINK_AFFILIATIONS');
export const updateRelatedArticleAction = createAction<RelatedArticle>('UPDATE_RELATED_ARTICLE');
export const addRelatedArticleAction = createAction<RelatedArticle>('ADD_RELATED_ARTICLE');
export const deleteRelatedArticleAction = createAction<RelatedArticle>('DELETE_RELATED_ARTICLE');
export const updateReferenceAction = createAction<Reference>('UPDATE_REFERENCE');
export const addReferenceAction = createAction<Reference>('ADD_REFERENCE');
export const deleteReferenceAction = createAction<Reference>('DELETE_REFERENCE');
export const applyChangeAction = createAction<ApplyChangePayload>('APPLY_CHANGE');

export const undoAction = createAction<void>('UNDO');
export const redoAction = createAction<void>('REDO');
export const toggleMarkAction = createAction<string>('TOGGLE_MARK');
export const insertReferenceCitationAction = createAction<void>('INSERT_REFERENCE_CITATION');
export const insertBoxAction = createAction<void>('INSERT_BOX');
export const insertListAction = createAction<ListType>('INSERT_LIST');
export const insertFigureAction = createAction<string>('INSERT_FIGURE');
export const insertFigureCitationAction = createAction<void>('INSERT_FIGURE_CITATION');
export const insertHeadingAction = createAction<number>('INSERT_HEADING');
export const insertParagraphAction = createAction<void>('INSERT_PARAGRAPH');
