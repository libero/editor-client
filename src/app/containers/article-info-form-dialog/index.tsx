import React, { useCallback, SyntheticEvent, useState, ChangeEvent } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import { isEqual, set } from 'lodash';
import Interweave from 'interweave';

import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { ActionButton } from 'app/components/action-button';
import { useArticleInfoFormStyles } from 'app/containers/article-info-form-dialog/styles';
import { getArticleInformation, getAuthors } from 'app/selectors/manuscript.selectors';
import formGrid from 'app/styles/form-grid.module.scss';
import { Select } from 'app/components/select';
import {
  ArticleInformation,
  getCopyrightStatement,
  getLicenseTextEditorState,
  LICENSE_CC0,
  LICENSE_CC_BY_4
} from 'app/models/article-information';
import { SectionContainer } from 'app/components/section-container';
import { stringifyEditorState } from 'app/utils/view.utils';

const labelProps = { shrink: true };

const SubjectOptions = [
  'Biochemistry and Chemical Biology',
  'Cancer Biology',
  'Cell Biology',
  'Chromosomes and Gene Expression',
  'Computational and Systems Biology',
  'Developmental Biology',
  'Ecology',
  'Epidemiology and Global Health',
  'Evolutionary Biology',
  'Genetics and Genomics',
  'Medicine',
  'Immunology and Inflammation',
  'Microbiology and Infectious Disease',
  'Neuroscience',
  'Physics of Living Systems',
  'Plant Biology',
  'Stem Cells and Regenerative Medicine',
  'Structural Biology and Molecular Biophysics'
].map((title) => ({ label: title, value: title }));

export const ArticleInfoFormDialog: React.FC<{}> = () => {
  const classes = useArticleInfoFormStyles();
  const dispatch = useDispatch();
  const articleInfo = useSelector(getArticleInformation);

  const [userArticleInfo, setArticleInfo] = useState<ArticleInformation>(articleInfo);
  const authors = useSelector(getAuthors);
  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      setArticleInfo({
        ...userArticleInfo,
        [fieldName]: newValue
      });
    },
    [userArticleInfo, setArticleInfo]
  );

  const handleSubjectChange = useCallback(
    (event: ChangeEvent<{ name: string; value: string }>) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      const updatedArticleInfo = {
        ...userArticleInfo,
        subjects: [...userArticleInfo.subjects]
      };

      set(updatedArticleInfo, fieldName, newValue);
      updatedArticleInfo.subjects = updatedArticleInfo.subjects.filter(Boolean);
      setArticleInfo(updatedArticleInfo);
    },
    [userArticleInfo, setArticleInfo]
  );

  const handleLicenseTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: string }>) => {
      const newLicenseType = event.target['value'];
      const updatedArticleInfo = {
        ...userArticleInfo,
        licenseType: newLicenseType,
        licenseText: getLicenseTextEditorState(newLicenseType),
        copyrightStatement:
          newLicenseType === LICENSE_CC_BY_4 ? getCopyrightStatement(authors, userArticleInfo.publicationDate) : ''
      };
      setArticleInfo(updatedArticleInfo);
    },
    [userArticleInfo, authors]
  );

  const handleAccept = useCallback(() => {
    if (!isEqual(userArticleInfo, articleInfo)) {
      dispatch(manuscriptActions.updateArticleInformationAction(userArticleInfo));
    }
    closeDialog();
  }, [closeDialog, articleInfo, userArticleInfo, dispatch]);

  return (
    <section className={classes.root}>
      <div className={formGrid.container}>
        <Select
          className={formGrid.fullWidth}
          name="subjects.0"
          placeholder="Please select"
          fullWidth
          canUnselect={true}
          blankValue={undefined}
          label="Subject *"
          value={userArticleInfo.subjects[0]}
          onChange={handleSubjectChange}
          options={SubjectOptions}
        />
        <Select
          className={formGrid.fullWidth}
          name="subjects.1"
          placeholder="Please select"
          fullWidth
          canUnselect={true}
          blankValue={undefined}
          label="Subject"
          value={userArticleInfo.subjects[1]}
          onChange={handleSubjectChange}
          options={SubjectOptions}
        />
        <TextField
          fullWidth
          name="articleDOI"
          label="Article DOI"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.articleDOI}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="publisherId"
          label="Publisher ID"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.publisherId}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="elocationId"
          label="eLocation ID"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.elocationId}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="volume"
          label="Volume"
          classes={{ root: formGrid.firstCol }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.volume}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="publicationDate"
          label="Published"
          type="date"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.publicationDate}
          onChange={handleFormChange}
        />
        <Select
          className={formGrid.fullWidth}
          placeholder="Please select"
          fullWidth
          blankValue={undefined}
          label="License type"
          value={userArticleInfo.licenseType}
          onChange={handleLicenseTypeChange}
          options={[
            { label: LICENSE_CC_BY_4, value: LICENSE_CC_BY_4 },
            { label: LICENSE_CC0, value: LICENSE_CC0 }
          ]}
        />
        {userArticleInfo.licenseType === LICENSE_CC_BY_4 ? (
          <SectionContainer label="Copyright statement" variant="outlined" className={formGrid.fullWidth}>
            {userArticleInfo.copyrightStatement}
          </SectionContainer>
        ) : undefined}
        <SectionContainer label="Permissions" variant="outlined" className={formGrid.fullWidth}>
          <Interweave content={stringifyEditorState(userArticleInfo.licenseText)} />
        </SectionContainer>
      </div>
      <div className={classes.buttonPanel}>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleAccept} title="Done" />
      </div>
    </section>
  );
};
