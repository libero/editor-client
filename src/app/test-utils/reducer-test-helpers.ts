import { EditorState } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';

import { Manuscript } from 'app/models/manuscript';
import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { ManuscriptHistoryState } from 'app/store';

export function givenState(overrides: Partial<Manuscript>): ManuscriptHistoryState {
  return getLoadableStateSuccess(
    getInitialHistory({
      title: createDummyEditorState(),
      abstract: createDummyEditorState(),
      authors: [],
      affiliations: [],
      keywordGroups: {},
      articleInfo: {
        dtd: '',
        articleType: '',
        articleDOI: ''
      },
      references: [],
      ...overrides
    })
  );
}

export function createDummyEditorState(): EditorState {
  return EditorState.create({
    schema: new Schema({
      nodes: {
        doc: { content: 'inline*' },
        text: { inline: true, group: 'inline' }
      }
    }),
    plugins: []
  });
}
