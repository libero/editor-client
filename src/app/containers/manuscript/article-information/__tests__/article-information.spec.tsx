import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import * as manuscriptEditorActions from '../../../../actions/manuscript-editor.actions';
import { givenState } from '../../../../test-utils/reducer-test-helpers';
import { ArticleInformation } from '../';
import { IconButton } from '@material-ui/core';
import { ArticleInfoFormDialog } from '../../../article-info-form-dialog';

jest.mock('@material-ui/core', () => ({
  Button: ({ label }) => <div data-cmp="Button">{label}</div>,
  IconButton: () => <div data-cmp="IconButton"></div>
}));

describe('Article information', () => {
  const mockStore = configureMockStore([]);

  it('renders article information', () => {
    const mockState = givenState({});
    Object.assign(mockState.data.present.articleInfo, {
      articleDOI: '',
      dtd: '1',
      articleType: '',
      publisherId: ''
    });
    const store = mockStore({
      manuscript: mockState
    });
    const wrapper = create(
      <Provider store={store}>
        <ArticleInformation />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('edits article information', () => {
    const mockState = givenState({});
    mockState.data.present.articleInfo.articleDOI = 'SOME_ID';
    const store = mockStore({
      manuscript: mockState
    });
    jest.spyOn(store, 'dispatch');
    const wrapper = mount(
      <Provider store={store}>
        <ArticleInformation />
      </Provider>
    );

    wrapper.find(IconButton).prop('onClick')(null);
    expect(store.dispatch).toHaveBeenCalledWith(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  });
});
