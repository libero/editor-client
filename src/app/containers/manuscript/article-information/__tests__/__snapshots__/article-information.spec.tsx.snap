// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article information renders article information 1`] = `
<div
  className="makeStyles-sectionContainer-1 makeStyles-root-4 makeStyles-plainVariant-6"
>
  <label
    className="makeStyles-label-7"
  >
    Article information
  </label>
  <section
    className="makeStyles-infoSection-2"
  >
    <div>
      <strong>
         Subject(s): 
      </strong>
       
      
    </div>
    <div>
      <strong>
         Article type: 
      </strong>
       
      
    </div>
    <div>
      <strong>
         Article DOI: 
      </strong>
       
      
    </div>
    <div>
      <strong>
         Publisher ID: 
      </strong>
       
      
    </div>
    <div>
      <strong>
         eLocation ID: 
      </strong>
       
      
    </div>
    <div>
      <strong>
         Year: 
      </strong>
       
      
    </div>
    <div>
      <strong>
         Volume: 
      </strong>
       
      
    </div>
    <div>
      Â 
    </div>
    <div>
      <strong>
         Published: 
      </strong>
       
      
    </div>
  </section>
  <div
    data-cmp="IconButton"
  />
</div>
`;
